AK/BufferedStream.h

```diff
@@ -115,6 +115,73 @@ class BufferedHelper {
         return m_buffer.read(buffer);
     }
 
+    ErrorOr<StringView> read_line_with_resize(ByteBuffer& buffer)
+    {
+        return StringView { TRY(read_until_with_resize(buffer, "\n"sv)) };
+    }
+
+    ErrorOr<Bytes> read_until_with_resize(ByteBuffer& buffer, StringView candidate)
+    {
+        return read_until_any_of_with_resize(buffer, Array { candidate });
+    }
+
+    template<size_t N>
+    ErrorOr<Bytes> read_until_any_of_with_resize(ByteBuffer& buffer, Array<StringView, N> candidates)
+    {
+        if (!stream().is_open())
+            return Error::from_errno(ENOTCONN);
+
+        if (buffer.is_empty())
+            return buffer;
+
+        auto candidate = TRY(find_and_populate_until_any_of(candidates));
+
+        if (candidate.has_value()) {
+            buffer.resize(candidate->offset);
+            auto const read_bytes = m_buffer.read(buffer);
+            TRY(m_buffer.discard(candidate->size));
+            return read_bytes;
+        }
+
+        if (m_buffer.used_space() == 0 && stream().is_eof()) {
+            // There are no bytes left in m_buffer or in the stream,
+            // return an empty buffer
+            buffer.resize(0);
+            return buffer.span();
+        }
+
+        if (!stream().is_eof()) {
+            // At this point we have completely filled m_buffer without finding a candidate
+            VERIFY(m_buffer.empty_space() == 0);
+        }
+        // Note: past this point the user supplied buffer will grow larger than m_buffer
+        size_t bytes_read_to_user_buffer = 0;
+        while (!candidate.has_value()) {
+            if (buffer.size() - bytes_read_to_user_buffer < m_buffer.capacity()) {
+                // m_buffer is full and buffer does not have enough space to hold its contents
+                // so we resize it.
+                buffer.resize((buffer.size() + m_buffer.capacity()) * 1.5);
```

Comment:
I'm not sure if we should try to do resize optimizations yet. For now, just saying "if we need to make room for `m_buffer.capacity()` bytes, then do exactly that" should be fine.